grpc/src/main/java/com/linecorp/armeria/common/grpc/protocol/AbstractUnaryGrpcService.java

```diff
@@ -0,0 +1,122 @@
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.grpc.GrpcHeaderNames;
+import com.linecorp.armeria.common.grpc.protocol.ArmeriaMessageDeframer.ByteBufOrStream;
+import com.linecorp.armeria.common.grpc.protocol.ArmeriaMessageDeframer.Listener;
+import com.linecorp.armeria.server.AbstractHttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+
+/**
+ * An {@link AbstractUnaryGrpcService} can be used to implement a gRPC service without depending on gRPC stubs.
+ * This service takes care of deframing and framing with the gRPC wire format and handling appropriate headers.
+ *
+ * <p>This service does not support compression. If you need support for compression, please consider using
+ * normal gRPC stubs or file a feature request.
+ */
+public abstract class AbstractUnaryGrpcService extends AbstractHttpService {
+
+    private static final HttpHeaders RESPONSE_HEADERS =
+            HttpHeaders.of(HttpStatus.OK)
+                       .add(GrpcHeaderNames.GRPC_ENCODING, "identity")
+                       .asImmutable();
+
+    /**
+     * Returns an unframed response message to return to the client, given an unframed request message. It is
+     * expected that the implementation has the logic to know how to parse the request and serialize a response
+     * into {@code byte[]}. The returned {@code byte[]} will be framed and returned to the client.
+     */
+    protected abstract CompletableFuture<byte[]> handleMessage(byte[] message);
+
+    @Override
+    protected final HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req) {
+        final CompletableFuture<HttpResponse> responseFuture =
+                req.aggregateWithPooledObjects(ctx.contextAwareEventLoop(), ctx.alloc())
+                   .thenCompose(msg -> deframeMessage(msg.content(), ctx.alloc()))
+                   .thenCompose(requestMessage -> {
+                       final byte[] requestMessageBytes;
+                       try {
+                           requestMessageBytes = ByteBufUtil.getBytes(requestMessage);
+                       } finally {
+                           requestMessage.release();
+                       }
+                       return handleMessage(requestMessageBytes);
+                   })
+                   .thenApply(responseMessage -> {
+                       final ArmeriaMessageFramer framer = new ArmeriaMessageFramer(
+                               ctx.alloc(), Integer.MAX_VALUE);
+                       final HttpData framed = framer.writePayload(Unpooled.wrappedBuffer(responseMessage));
+                       return HttpResponse.of(
```

Comment:
this is where my empty response gets.. though the content is different and that trips up gRPC client (might be my mistake)